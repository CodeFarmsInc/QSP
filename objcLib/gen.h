#ifndef ZZ_CODEGEN_INCLUDE
#define ZZ_CODEGEN_INCLUDE

#include "lib/general.h"

class Generator;
class Organization;
class Participant;
class OrgType;
class PartType;
class Iterator;
class FixedType;
class ClassLink;
class ApplClass;

#include "tmp/orgs.h"
#include "tmp/orgs1.h"
#include "tmp/parts.h"
#include "tmp/parts1.h"
#include "tmp/parts2.h"
#include "tmp/orgName.h"
#include "tmp/partName.h"
#include "tmp/types.h"
#include "tmp/types1.h"
#include "tmp/partTypes.h"
#include "tmp/partTypes1.h"
#include "tmp/partTypes2.h"
#include "tmp/inherits.h"
#include "tmp/iters.h"
#include "tmp/iters1.h"
#include "tmp/orgTypeName.h"
#include "tmp/fileName.h"
#include "tmp/partTypeName.h"
#include "tmp/iterName.h"
#include "tmp/inhName.h"
#include "tmp/fixedName.h"
#include "tmp/inhMatch.h"
#include "tmp/inhFixed.h"
#include "tmp/fixedTypes.h"
#include "tmp/fixedTypes1.h"
#include "tmp/myType.h"
#include "tmp/myPartType.h"
#include "tmp/toBase.h"
#include "tmp/classes.h"
#include "tmp/classes1.h"
#include "tmp/dependsOn.h"
#include "tmp/dependsOn1.h"
#include "tmp/usedBy.h"
#include "tmp/usedBy1.h"
#include "tmp/usedBy2.h"
#include "tmp/cName.h"

// -----------------------------------------------
typedef class orgs_LinkedList2 orgs;
typedef class parts_Aggregate2 parts;
typedef class orgName_Name orgName;
typedef class partName_Name partName;
typedef class types_LinkedList2 types;
typedef class partTypes_Aggregate2 partTypes;
typedef class inherits_SingleLink inherits;
typedef class iters_LinkedList1 iters;
typedef class orgTypeName_Name orgTypeName;
typedef class fileName_Name fileName;
typedef class partTypeName_Name partTypeName;
typedef class iterName_Name iterName;
typedef class inhName_Name inhName;
typedef class fixedName_Name fixedName;
typedef class inhMatch_SingleLink inhMatch;
typedef class inhFixed_SingleLink inhFixed;
typedef class fixedTypes_LinkedList1 fixedTypes;
typedef class myType_SingleLink myType;
typedef class myPartType_SingleLink myPartType;
typedef class toBase_SingleLink toBase;
typedef class classes_Hash classes;
typedef class dependsOn_LinkedList1 dependsOn;
typedef class usedBy_Aggregate2 usedBy;
typedef class cName_Name cName;
// -----------------------------------------------


class ZZ_Generator {
public:
    orgs_LinkedList2Parent ZZorgs;
    types_LinkedList2Parent ZZtypes;
    classes_HashHolder ZZclasses;
};

class ZZ_Organization {
public:
    orgs_LinkedList2Child ZZorgs;
    parts_Aggregate2Parent ZZparts;
    orgName_NameParent ZZorgName;
    myType_SingleLinkParent ZZmyType;
};

class ZZ_Participant {
public:
    parts_Aggregate2Child ZZparts;
    partName_NameParent ZZpartName;
    myPartType_SingleLinkParent ZZmyPartType;
    usedBy_Aggregate2Child ZZusedBy;
};

class ZZ_OrgType {
public:
    types_LinkedList2Child ZZtypes;
    partTypes_Aggregate2Parent ZZpartTypes;
    inherits_SingleLinkParent ZZinherits;
    iters_LinkedList1Parent ZZiters;
    orgTypeName_NameParent ZZorgTypeName;
    fileName_NameParent ZZfileName;
    fixedTypes_LinkedList1Parent ZZfixedTypes;
};

class ZZ_PartType {
public:
    partTypes_Aggregate2Child ZZpartTypes;
    partTypeName_NameParent ZZpartTypeName;
    inhName_NameParent ZZinhName;
    inhMatch_SingleLinkParent ZZinhMatch;
};

class ZZ_Iterator {
public:
    iters_LinkedList1Child ZZiters;
    iterName_NameParent ZZiterName;
};

class ZZ_FixedType {
public:
    fixedName_NameParent ZZfixedName;
    inhFixed_SingleLinkParent ZZinhFixed;
    fixedTypes_LinkedList1Child ZZfixedTypes;
};

class ZZ_ClassLink {
public:
    toBase_SingleLinkParent ZZtoBase;
    dependsOn_LinkedList1Child ZZdependsOn;
};

class ZZ_ApplClass {
public:
    classes_HashElement ZZclasses;
    dependsOn_LinkedList1Parent ZZdependsOn;
    usedBy_Aggregate2Parent ZZusedBy;
    cName_NameParent ZZcName;
};

typedef orgs_LinkedList2Iterator orgs_Iterator;
typedef parts_Aggregate2Iterator parts_Iterator;
typedef types_LinkedList2Iterator types_Iterator;
typedef partTypes_Aggregate2Iterator partTypes_Iterator;
typedef iters_LinkedList1Iterator iters_Iterator;
typedef fixedTypes_LinkedList1Iterator fixedTypes_Iterator;
typedef classes_HashIterator classes_Iterator;
typedef dependsOn_LinkedList1Iterator dependsOn_Iterator;
typedef usedBy_Aggregate2Iterator usedBy_Iterator;

#endif // ZZ_CODEGEN_INCLUDE
