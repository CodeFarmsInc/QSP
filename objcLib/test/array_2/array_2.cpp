#include "array_2.h"

#define BSIZE 80

main(int argc,char **argv) {
    int pSort(const void *p1,const void *p2);
    int oSort(const void *p1,const void *p2);
    int hSort(const void *p1,const void *p2);
    void cpEmpl(Employee *p1,Employee *p2),freeArr(Header *h);
    void bck(void *,int),hBck(void *,int);
    FILE *file1,*file2;
    Header *h;
    Employee *ep,*sp,**pa,*oa,*p1,**p2,*p3,*p4,*ee;
    PtrEmpl *pe,*pp;
    char buff[BSIZE],name[BSIZE],*cp,*cc;
    int i,k,n;

    if(argc<=1)file1=stdin;  else file1=fopen(argv[1],"r");
    if(argc<=2)file2=stdout; else file2=fopen(argv[2],"w");

    h=new Header;    /* main holder of the arrays */
    ee=new Employee; /* temporary working object */
    pp=new PtrEmpl;  /* temporary working object */
    sp=(Employee *)NULL;
    n=0;
    while(fgets(buff,BSIZE,file1)){
        ep=new Employee;
        sscanf(buff,"%s %d %d\n",name,&(ep->emplNum),&(ep->salary));
        cp=new char[strlen(name)+1]; strcpy(cp,name);
        eName.add(ep,cp);
        sp=eRing.add(sp,ep);
        n++;
    }
    rStart.add(h,sp);
    fclose(file1);

    /* test of normal usage of the array, compare all 4 methods */
    fprintf(file2," testing 4 ways to index the array\n");
    pa=pArr.form(h,4,2);
    oa=oArr.form(h,3,-2);

    i=0;
    eRing_iterator eIter(sp);
    for(ep= ++eIter; ep; ep= ++eIter){
        p2=pArr.ind(h,i);
        (*p2)=ep;
        p1=oArr.ind(h,n-i);
        cpEmpl(ep,p1);
        i++;
    }

    /* update array entries before section of direct access */
    pa=pArr.ind(h,0);
    oa=oArr.ind(h,0);
    fprintf(file2,"\n direct access\n");
    for(i=0;i<n;i++){
        p1=pa[i];
        cp=eName.fwd(p1);
        fprintf(file2,"i=%d name=%s emplNum=%d salary=%d",
                                               i,cp,p1->emplNum,p1->salary);
        p3= &(oa[n-i]);
        cc=eName.fwd(p3);
        if(strcmp(cp,cc))fprintf(file2," (1 not 3)");
        fprintf(file2,"\n");
    }
    fprintf(file2,"\n protected access\n");
    for(i=0;i<n;i++){
        p2=pArr.ind(h,i);
        cp=eName.fwd(*p2);
        fprintf(file2,"i=%d name=%s emplNum=%d salary=%d",
                                            i,cp,(*p2)->emplNum,(*p2)->salary);
        p4=oArr.ind(h,n-i);
        cc=eName.fwd(p4);
        if(strcmp(cp,cc))fprintf(file2," (2 not 4)");
        fprintf(file2,"\n");
    }

    fprintf(file2,"\n the same thing, only different indexing\n");
    freeArr(h);
    pa=pArr.form(h,4,2);
    oa=oArr.form(h,3,-2);

    i=0;
    eIter.start(sp);
    for(ep= ++eIter; ep; ep= ++ eIter){
        p2=pArr.ind(h,n-i);
        (*p2)=ep;
        p1=oArr.ind(h,i);
        cpEmpl(ep,p1);
        i++;
    }

    /* update array entries before section of direct access */
    pa=pArr.ind(h,0);
    oa=oArr.ind(h,0);
    fprintf(file2,"\n direct access\n");
    for(i=0;i<n;i++){
        p1=pa[n-i];
        cp=eName.fwd(p1);
        fprintf(file2,"i=%d name=%s emplNum=%d salary=%d",
                                               i,cp,p1->emplNum,p1->salary);
        p3= &(oa[i]);
        cc=eName.fwd(p3);
        if(strcmp(cp,cc))fprintf(file2," (1 not 3)");
        fprintf(file2,"\n");
    }
    fprintf(file2,"\n protected access\n");
    for(i=0;i<n;i++){
        p2=pArr.ind(h,n-i);
        cp=eName.fwd(*p2);
        fprintf(file2,"i=%d name=%s emplNum=%d salary=%d",
                                            i,cp,(*p2)->emplNum,(*p2)->salary);
        p4=oArr.ind(h,i);
        cc=eName.fwd(p4);
        if(strcmp(cp,cc))fprintf(file2," (2 not 4)");
        fprintf(file2,"\n");
    }

    fprintf(file2,"\n using the array as a stack\n");
    freeArr(h);
    pa=pArr.form(h,4,-3);
    oa=oArr.form(h,3,10);

    eIter.start(sp);
    for(ep= ++eIter; ep; ep= ++ eIter){
        p2= &ep;
        pArr.push(h,p2);
        cpEmpl(ep,ee);
        oArr.push(h,ee);
        cp=eName.fwd(ep);
        fprintf(file2,"push: %s emplNum=%d salary=%d\n",
                                                 cp,ep->emplNum,ep->salary);
    }
    for(p1=sp;p1;){ /* start with any non-null pointer from loading eRing */
        p2=pArr.pop(h);
        if(!p2)break;
        cc=eName.fwd(*p2);
        p1=oArr.pop(h);
        if(!p1)break;
        cp=eName.fwd(p1);
        fprintf(file2,"pop: %s emplNum=%d salary=%d",
                                                 cp,p1->emplNum,p1->salary);
        if(!strcmp(cc,cp))fprintf(file2,"\n");
        else fprintf(file2," different names: %s %s\n",cc,cp);
    }

    fprintf(file2,"\n using the array as a heap, load twice\n");
    freeArr(h);
    pa=pArr.form(h,4,-3);
    oa=oArr.form(h,3,3);

    /* load the heap */
    eIter.start(sp);
    for(ep= ++eIter; ep; ep= ++ eIter){
        p2= &ep;
        pArr.inHeap(pSort,h,p2,bck);
        cpEmpl(ep,ee);
        oArr.inHeap(oSort,h,ee,bck);
        cp=eName.fwd(ep);
        fprintf(file2,"in heap: %s emplNum=%d salary=%d\n",
                                                 cp,ep->emplNum,ep->salary);
    }
    /* three hand modifications */
    pArr.delHeap(pSort,h,4,bck);
    oArr.delHeap(oSort,h,4,bck);
    p2=pArr.ind(h,4);
    p1=oArr.ind(h,4);
    p1->salary=(*p2)->salary=4200;
    pArr.updHeap(pSort,h,4,bck);
    oArr.updHeap(oSort,h,4,bck);
    p2=pArr.ind(h,5);
    p1=oArr.ind(h,5);
    p1->salary=(*p2)->salary=1400;
    pArr.updHeap(pSort,h,5,bck);
    oArr.updHeap(oSort,h,5,bck);

    /* unload the heap */
    /* p1 and ee are the storage for the intermediate results */
    for(;;){ /* end of the loop detected by the id.size() function */
        n=pArr.size(h);
        if(n<=0)break;
        pArr.outHeap(pSort,h,&p1,bck);
        cc=eName.fwd(p1);
        n=oArr.size(h);
        if(n<=0)break;
        oArr.outHeap(oSort,h,ee,bck); /* ee points to an allocated object */
        cp=eName.fwd(ee);
        fprintf(file2,"out heap: %s emplNum=%d salary=%d",
                                                 cp,p1->emplNum,ee->salary);
        if(!strcmp(cc,cp))fprintf(file2,"\n");
        else fprintf(file2," different names: %s %s\n",cc,cp);
    }

    fprintf(file2,"\n heap with a call-back function\n");
    freeArr(h);
    pe=hArr.form(h,4,-3);

    /* load the heap */
    eIter.start(sp);
    ITERATE(eIter,ep){
        pe=new PtrEmpl;
        hPtr.add(pe,ep);
        hArr.inHeap(hSort,h,pe,hBck);
        cp=eName.fwd(ep);
        fprintf(file2,"in heap: %s emplNum=%d salary=%d\n",
                                                 cp,ep->emplNum,ep->salary);
    }
    /* delete odd entries, modify salaries +-1000 */
    i=1;
    n=1000;
    eIter.start(sp);
    ITERATE(eIter,ep){
        k=ep->heapIndex;
        cp=eName.fwd(ep);
        if(i>0){
            hArr.delHeap(hSort,h,k,hBck);
            fprintf(file2,"deleted: %s emplNum=%d salary=%d\n",
                                          cp,ep->emplNum,ep->salary);
        }
        else {
            ep->salary+=n;
            n= -n;
            hArr.updHeap(hSort,h,k,hBck);
            fprintf(file2,"update: %s emplNum=%d salary=%d\n",
                                          cp,ep->emplNum,ep->salary);
        }
        i= -i;
    };
    /* unload the heap */
    /* pp is a pre-allocated storage for intermediate results */
    for(;;){ /* end of the loop detected by pArr.size() */
        n=hArr.size(h);
        if(n<=0)break;
        hArr.outHeap(hSort,h,pp,bck);
        ep=hPtr.target(pp);
        cp=eName.fwd(ep);
        fprintf(file2,"out heap: %s emplNum=%d salary=%d\n",
                                                 cp,ep->emplNum,ep->salary);
    }
    return(0);
}
/* --------------------------------------------------------------------- */
int oSort(const void *c1,const void *c2)
{
    Employee *p1,*p2;

    p1=(Employee *)c1;
    p2=(Employee *)c2;
    if(p1->salary<p2->salary)return(-1);
    if(p1->salary>p2->salary)return(1);
    return(0);
}
/* --------------------------------------------------------------------- */
int pSort(const void *c1,const void *c2)
{
    Employee **p1,**p2;

    p1=(Employee **)c1;
    p2=(Employee **)c2;
    if((*p1)->salary<(*p2)->salary)return(-1);
    if((*p1)->salary>(*p2)->salary)return(1);
    return(0);
}
/* --------------------------------------------------------------------- */
int hSort(const void *c1,const void *c2)
{
    Employee *p1,*p2;

    p1=hPtr.target((PtrEmpl*)c1);
    p2=hPtr.target((PtrEmpl*)c2);
    if(p1->salary<p2->salary)return(-1);
    if(p1->salary>p2->salary)return(1);
    return(0);
}
/* --------------------------------------------------------------------- */
/* disconnect links on array oa, and free both arrays */
void freeArr(Header *h)
{
    int i,n;
    Employee *oa;

    n=oArr.capacity(h);
    oa=oArr.ind(h,0);
    for(i=0;i<n;i++){
        eName.del(&(oa[i]));
    }
    pArr.free(h);
    oArr.free(h);
    return;
}
/* --------------------------------------------------------------------- */
/* Copy one Employee record onto another, but avoid the organization pointers.
 * Note that: (*p2)=(*p1); would be wrong, it would mess up the pointers */
void cpEmpl(Employee *p1,Employee *p2)
{
    char *cp;

    eName.del(p2);
    cp=eName.fwd(p1);
    eName.add(p2,cp); /* using the same name as string */
    p2->emplNum=p1->emplNum;
    p2->salary=p1->salary;
    return;
}
// --------------------------------------------------------------------- */
// heap callback function when not used
void bck(void *p,int i){p=p; i=i;}

// heap callback function when used
void hBck(void *p,int i){
    Employee *e;

    e=hPtr.target((PtrEmpl*)p);
    e->heapIndex=i;
}

#include "gen.cpp"
