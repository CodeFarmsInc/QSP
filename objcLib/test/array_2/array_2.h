
// ------------------------------------------------------
// This is test9a from the DOL regression suite, and the results
// should be identical. 
// This is an extensive test of the Array class, together with Ring1,
// SingleLink and Name organizations.

// The Array class is tested as a basic array, array of pointers, binary
// heap, and stack.

// The test problem:
// A set of employee records is loaded into a ring (eRing).
// and from there into array, stack, and heap (hArr). 
// Two sets of tests are made: one for the array of full employee records(oArr),
// the other for the array of pointers (pArr).
// All arrays start with a smaller size than actually
// needed, so that the re-allocation mechanism is tested.

// Input format: employeeName id salary
// ------------------------------------------------------

#include <stdio.h>
#include <string.h>
#include "gen.h"

class Header {
public:
    ZZ_Header ZZds;
};
class Employee {
public:
    ZZ_Employee ZZds;
    int emplNum; // employee number or id
    int salary;
    int heapIndex;
};
class PtrEmpl {
public:
    ZZ_PtrEmpl ZZds;
};

