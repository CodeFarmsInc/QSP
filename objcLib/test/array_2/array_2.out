 testing 4 ways to index the array

 direct access
i=0 name=T.Seal emplNum=2901 salary=2400
i=1 name=W.Water emplNum=1100 salary=2400
i=2 name=E.Light emplNum=3387 salary=2045
i=3 name=J.Shade emplNum=1987 salary=1900
i=4 name=L.Dotty emplNum=7543 salary=3230
i=5 name=F.Gold emplNum=7990 salary=3905
i=6 name=P.Grey emplNum=1289 salary=4140
i=7 name=S.White emplNum=3523 salary=2400
i=8 name=I.Black emplNum=7003 salary=2795
i=9 name=A.Green emplNum=1876 salary=3500
i=10 name=J.Brown emplNum=4367 salary=2105

 protected access
i=0 name=T.Seal emplNum=2901 salary=2400
i=1 name=W.Water emplNum=1100 salary=2400
i=2 name=E.Light emplNum=3387 salary=2045
i=3 name=J.Shade emplNum=1987 salary=1900
i=4 name=L.Dotty emplNum=7543 salary=3230
i=5 name=F.Gold emplNum=7990 salary=3905
i=6 name=P.Grey emplNum=1289 salary=4140
i=7 name=S.White emplNum=3523 salary=2400
i=8 name=I.Black emplNum=7003 salary=2795
i=9 name=A.Green emplNum=1876 salary=3500
i=10 name=J.Brown emplNum=4367 salary=2105

 the same thing, only different indexing

 direct access
i=0 name=T.Seal emplNum=2901 salary=2400
i=1 name=W.Water emplNum=1100 salary=2400
i=2 name=E.Light emplNum=3387 salary=2045
i=3 name=J.Shade emplNum=1987 salary=1900
i=4 name=L.Dotty emplNum=7543 salary=3230
i=5 name=F.Gold emplNum=7990 salary=3905
i=6 name=P.Grey emplNum=1289 salary=4140
i=7 name=S.White emplNum=3523 salary=2400
i=8 name=I.Black emplNum=7003 salary=2795
i=9 name=A.Green emplNum=1876 salary=3500
i=10 name=J.Brown emplNum=4367 salary=2105

 protected access
i=0 name=T.Seal emplNum=2901 salary=2400
i=1 name=W.Water emplNum=1100 salary=2400
i=2 name=E.Light emplNum=3387 salary=2045
i=3 name=J.Shade emplNum=1987 salary=1900
i=4 name=L.Dotty emplNum=7543 salary=3230
i=5 name=F.Gold emplNum=7990 salary=3905
i=6 name=P.Grey emplNum=1289 salary=4140
i=7 name=S.White emplNum=3523 salary=2400
i=8 name=I.Black emplNum=7003 salary=2795
i=9 name=A.Green emplNum=1876 salary=3500
i=10 name=J.Brown emplNum=4367 salary=2105

 using the array as a stack
push: T.Seal emplNum=2901 salary=2400
push: W.Water emplNum=1100 salary=2400
push: E.Light emplNum=3387 salary=2045
push: J.Shade emplNum=1987 salary=1900
push: L.Dotty emplNum=7543 salary=3230
push: F.Gold emplNum=7990 salary=3905
push: P.Grey emplNum=1289 salary=4140
push: S.White emplNum=3523 salary=2400
push: I.Black emplNum=7003 salary=2795
push: A.Green emplNum=1876 salary=3500
push: J.Brown emplNum=4367 salary=2105
pop: J.Brown emplNum=4367 salary=2105
pop: A.Green emplNum=1876 salary=3500
pop: I.Black emplNum=7003 salary=2795
pop: S.White emplNum=3523 salary=2400
pop: P.Grey emplNum=1289 salary=4140
pop: F.Gold emplNum=7990 salary=3905
pop: L.Dotty emplNum=7543 salary=3230
pop: J.Shade emplNum=1987 salary=1900
pop: E.Light emplNum=3387 salary=2045
pop: W.Water emplNum=1100 salary=2400
pop: T.Seal emplNum=2901 salary=2400

 using the array as a heap, load twice
in heap: T.Seal emplNum=2901 salary=2400
in heap: W.Water emplNum=1100 salary=2400
in heap: E.Light emplNum=3387 salary=2045
in heap: J.Shade emplNum=1987 salary=1900
in heap: L.Dotty emplNum=7543 salary=3230
in heap: F.Gold emplNum=7990 salary=3905
in heap: P.Grey emplNum=1289 salary=4140
in heap: S.White emplNum=3523 salary=2400
in heap: I.Black emplNum=7003 salary=2795
in heap: A.Green emplNum=1876 salary=3500
in heap: J.Brown emplNum=4367 salary=2105
out heap: F.Gold emplNum=7990 salary=1400
out heap: J.Shade emplNum=1987 salary=1900
out heap: E.Light emplNum=3387 salary=2045
out heap: W.Water emplNum=1100 salary=2400
out heap: S.White emplNum=3523 salary=2400
out heap: T.Seal emplNum=2901 salary=2400
out heap: I.Black emplNum=7003 salary=2795
out heap: A.Green emplNum=1876 salary=3500
out heap: P.Grey emplNum=1289 salary=4140
out heap: L.Dotty emplNum=7543 salary=4200

 heap with a call-back function
in heap: T.Seal emplNum=2901 salary=2400
in heap: W.Water emplNum=1100 salary=2400
in heap: E.Light emplNum=3387 salary=2045
in heap: J.Shade emplNum=1987 salary=1900
in heap: L.Dotty emplNum=7543 salary=4200
in heap: F.Gold emplNum=7990 salary=1400
in heap: P.Grey emplNum=1289 salary=4140
in heap: S.White emplNum=3523 salary=2400
in heap: I.Black emplNum=7003 salary=2795
in heap: A.Green emplNum=1876 salary=3500
in heap: J.Brown emplNum=4367 salary=2105
deleted: T.Seal emplNum=2901 salary=2400
update: W.Water emplNum=1100 salary=3400
deleted: E.Light emplNum=3387 salary=2045
update: J.Shade emplNum=1987 salary=900
deleted: L.Dotty emplNum=7543 salary=4200
update: F.Gold emplNum=7990 salary=2400
deleted: P.Grey emplNum=1289 salary=4140
update: S.White emplNum=3523 salary=1400
deleted: I.Black emplNum=7003 salary=2795
update: A.Green emplNum=1876 salary=4500
deleted: J.Brown emplNum=4367 salary=2105
out heap: J.Shade emplNum=1987 salary=900
out heap: S.White emplNum=3523 salary=1400
out heap: F.Gold emplNum=7990 salary=2400
out heap: W.Water emplNum=1100 salary=3400
out heap: A.Green emplNum=1876 salary=4500
